packages/vm/src/context.rs

```diff
@@ -62,229 +58,243 @@ impl GasState {
+            gas_state: GasState::with_limit(gas_limit),
+            storage: None,
+            storage_readonly: true,
+            querier: None,
+            wasmer_instance: None,
+        }
     }
 }
 
 /// Returns the original storage and querier as owned instances, and closes any remaining
 /// iterators. This is meant to be called when recycling the instance.
 pub(crate) fn move_out_of_context<S: Storage, Q: Querier>(
-    source: &mut Ctx,
+    env: &Env<S, Q>,
 ) -> (Option<S>, Option<Q>) {
-    let b = get_context_data_mut::<S, Q>(source);
-    (b.storage.take(), b.querier.take())
+    env.with_context_data_mut(|context_data| {
+        (context_data.storage.take(), context_data.querier.take())
+    })
 }
 
-/// Moves owned instances of storage and querier into the context.
+/// Moves owned instances of storage and querier into the env.
 /// Should be followed by exactly one call to move_out_of_context when the instance is finished.
-pub(crate) fn move_into_context<S: Storage, Q: Querier>(target: &mut Ctx, storage: S, querier: Q) {
-    let b = get_context_data_mut::<S, Q>(target);
-    b.storage = Some(storage);
-    b.querier = Some(querier);
-}
-
-pub fn get_gas_state_mut<'a, 'b, S: Storage + 'b, Q: Querier + 'b>(
-    ctx: &'a mut Ctx,
-) -> &'b mut GasState {
-    &mut get_context_data_mut::<S, Q>(ctx).gas_state
-}
-
-pub fn get_gas_state<'a, 'b, S: Storage + 'b, Q: Querier + 'b>(ctx: &'a Ctx) -> &'b GasState {
-    &get_context_data::<S, Q>(ctx).gas_state
+pub(crate) fn move_into_context<S: Storage, Q: Querier>(env: &Env<S, Q>, storage: S, querier: Q) {
+    env.with_context_data_mut(|context_data| {
+        context_data.storage = Some(storage);
+        context_data.querier = Some(querier);
+    });
 }
 
-pub fn process_gas_info<S: Storage, Q: Querier>(ctx: &mut Ctx, info: GasInfo) -> VmResult<()> {
-    decrease_gas_left(ctx, info.cost)?;
-    account_for_externally_used_gas::<S, Q>(ctx, info.externally_used)?;
+pub fn process_gas_info<S: Storage, Q: Querier>(env: &Env<S, Q>, info: GasInfo) -> VmResult<()> {
+    decrease_gas_left(env, info.cost)?;
+    account_for_externally_used_gas(env, info.externally_used)?;
     Ok(())
 }
 
 /// Use this function to adjust the VM's gas limit when a call into the backend
 /// reported there was externally metered gas used.
 /// This does not increase the VM's gas usage but ensures the overall limit is not exceeded.
 fn account_for_externally_used_gas<S: Storage, Q: Querier>(
-    ctx: &mut Ctx,
+    env: &Env<S, Q>,
     amount: u64,
 ) -> VmResult<()> {
-    account_for_externally_used_gas_impl::<S, Q>(ctx, amount)
+    account_for_externally_used_gas_impl(env, amount)
 }
 
-#[cfg(feature = "default-singlepass")]
 fn account_for_externally_used_gas_impl<S: Storage, Q: Querier>(
-    ctx: &mut Ctx,
+    env: &Env<S, Q>,
     used_gas: u64,
 ) -> VmResult<()> {
-    use crate::backends::{get_gas_left, set_gas_left};
+    // WFT?!
+    let env1 = env.clone();
```

Comment:
I see why you need a clone here.. but really?

why not just 

```rust
let env2 = env.clone();
env.with_context_data_mut(...) {
  // replace `&env3` with `&env2`
}
```

Reply:
Or is the case of a read/write lock not being released properly?
And if clone() fixes that, it seems dangerous

Reply:
Wait... so we get a write lock on `env1` with the closure, right?
And then need a read lock on `env2` and write lock on `env3`?

I guess it is good you got this to work, but we should flag this section, it does seem dangerous (multiple refs to same object)

Reply:
Closing the envs is safe, but this could indeed trigger a panic when `data: Arc<RwLock<ContextData<S, Q>>>` is accessed the wrong way. Right now, `get_gas_left(&env2)` and `set_gas_left(&env3, new_limit)` are noops. Let's see when adding metering again.